olive/data/component/text_generation.py

```diff
@@ -79,10 +79,20 @@ def _check_padding(cls, v, values):
 class TextGenCorpusParams(TextGenParams):
     """Parameters for text generation task with 'corpus' dataset type."""
 
-    text_cols: list  # list of text columns
+    # list of text columns, for each row, columns are concatenated together using template
+    # for join corpus_strategy, all rows are then concatenated together using joiner
+    text_cols: list
+    # a python f-string template for the text, the text in each column is replaced with {column_name}
+    # is not provided, the text in each column is joined with a space
+    text_template: str = None
```

Comment:
>text_template [](http://example.com/codeflow?start=4&length=13)

a custom function might be better than f-string template?

Reply:
I plan to support custom formatting function in the future. But that needs more things like user script module and modifying the data container. Since all of these are absorbed in the kwargs, I think it will be okay to support multiple ways:
- column names
- fstring template
- formatting func

Reply:
could we add TODO there?

Reply:
added!

Reply:
BTW, I don't see the example use the template. are there any code snippet to demonstrate how they are used?

Reply:
There is no example in the code yet but I have a local example 
```
data_config = {
    "name": "tiny-code-python",
    "type": "HuggingfaceContainer",
    "params_config": {
        "model_name": model_name,
        "task": "text-generation",
        "data_name": "parquet",
        "data_files": "tiny-codes-python.parquet",
        "split": "train",
        "component_kwargs": {
            "pre_process_data": {
                "dataset_type": "corpus",
                "corpus_strategy": "join",
                "text_template": "### Question: {prompt} \n### Answer: {response}", 
                "max_samples": 1000,
                "source_max_len": 1024,
            }
        }
    }
}
```
It is not complete yet so I am waiting to add some other changes before adding this example.

Reply:
could we postpone the change for template into your examples so the reviewer or other guys would have the knowledge when looking at the commit history in future? Or we need add examples in this PR. 

In this way, we can guarantee the feature is completed when it is merged into main.

Reply:
the context bug could be a separated PR.

Reply:
I will add the example in this PR then. I need this PR to get merged so I can make follow up changes. Otherwise, the PR becomes too big.

Reply:
Basically, each PR should be as minimal as possible but keep it complete. At the same time, every PR should not include other unrelated changes. For example, the bug fix in this PR.

Reply:
Sometimes, it's hard to include a complete example since the full example depends on other features which in turn depends on the current PR. 
It's easier if we have unit tests for the components but currently don't have them for dataset components. 

For this PR I will add an example with a custom component that will be updated later to remove the custom component once that feature is added.
I will also create a separate PR for the bug-fix.

Reply:
as an opposite example, I have no idea about what the context bug it is since the file include a lot of other changes unless I carefully look at every line changes.

Reply:
Separate PR for context length https://github.com/microsoft/Olive/pull/581

Reply:
> Sometimes, it's hard to include a complete example since the full example depends on other features which in turn depends on the current PR. It's easier if we have unit tests for the components but currently don't have them for dataset components.
> 
> For this PR I will add an example with a custom component that will be updated later to remove the custom component once that feature is added. I will also create a separate PR for the bug-fix.

If there is no real dataset, an unit test should be enough by mocking up the dataset.

Reply:
Added example. Will rebase with main when the bugfix PR merges.

Reply:
rebased to main